### DISCUSSION TITLE
option to generate project with preact

### DISCUSSION THREAD

**Discussion by theArianit** [exhibitionist-digital/ultra#165] [Ideas]
Hello people.
First of all I want to praise you for the wonderful work for bringing SPA with react to deno.
But I am wondering if it would be possible (and hopefully not hard work) to add an option to choose preact instead of react. The library size doesn't bother me much, because in a real life project, the size difference would decrease with the complexity of the app (I am not sure, I am not an expert). But still, a 25% lighter js bundle can be expected with preact in comparison to react. Also the performance of preact seems to be a bit better in overall (according to some youtube videos and this [article](https://justinnoel.dev/2020/05/12/using-preact-in-a-next-js-project/) and this [one](https://javascript.plainenglish.io/i-built-the-same-app-with-react-and-preact-here-are-the-differences-b0da382a6f72).
Thanks for hearing out my suggestion

**Comment by deckchairlabs** [comment]
Hi @theArianit do you mean add preact to the `create` script? We have support for using `preact` instead of `react` as demonstrated here https://github.com/exhibitionist-digital/ultra/tree/main/examples/with-preact.

While using `preact` is possible, you do lose the benefits of Streaming, and we are unsure how far we will/can support `preact`

**Reply by theArianit** [reply]
Yes, that is basically what I meant. Thank for sharing the demo.
A noob question: what is the purpose of client.tsx and server.tsx? Sorry for this trivial question but I am trying to figure out the basics of these things

**Reply by deckchairlabs** [reply]
Not a problem! Our docs need serious work...

- `client.tsx` is for `browser/client` side hydration https://reactjs.org/docs/react-dom-client.html
- `server.tsx` renders your app/project on the server into HTML and sends that during an initial request.

**Reply by theArianit** [reply]
Oh, makes sense, and was expecting that answer. But for one moment I thought that the `server.ts` was for back-end.
I will keep both eyes on the documentation.
Glad that I found a full front-end SPA framework for deno. The internet is full with all the SSR, MPA, SSG just because they don't want the browser to not do any work at all. Keep up the good work

**Reply by deckchairlabs** [reply]
We're trying to keep it as flexible as possible, and not lock you into any kind of architecture, we loathe that. Feel free to open issues regarding docs or questions you might have, will force us to start writing seriously.

**Reply by theArianit** [reply]
You will def. hear more from me :P
Thanks for your help

**Reply by theArianit** [reply]
Hello @deckchairlabs 
about the preact example. wouldn't it be better to try to use the preact API when available, instead of react ones, like the case of client hydration?
I don't know what the roadmap is, but just giving my 2 cents

**Reply by deckchairlabs** [reply]
Do you mean within `client.tsx`? It isn't actually using the React API, since we are "aliasing" the `react` import to `https://esm.sh/preact@10.11.0/compat` which then uses Preact API for everything.

**Reply by deckchairlabs** [reply]
Note: pushed a commit which fixed an issue on with-preact example https://github.com/exhibitionist-digital/ultra/commit/2e9c73676a8b0c79ba052434f4d8d1b0788535a7

**Reply by deckchairlabs** [reply]
I suppose it would make sense to update that example to this instead.

```ts
import { hydrate } from "preact";
import App from "./src/app.tsx";

hydrate(<App />, document);
```

**Reply by theArianit** [reply]
yes, that's the example I meant.
But isn't `.../compat` the react API? I am sure that it is for react API compatibility

**Reply by deckchairlabs** [reply]
Correct, but under the hood is using Preact API

**Reply by theArianit** [reply]
> While using `preact` is possible, you do lose the benefits of Streaming, and we are unsure how far we will/can support `preact`

Sorry for coming back to this discussion after many days. I was trying to learn things under the hood but I can't figure out what you mean with streaming, and with the little time in disposal it's harder to. can you enlighten me here a bit?
I found another jsx alternative to react and preact: inferno js. and I am interested to add it as example if possible and if I figure out how to do things.
and since I am not a fan of using discord (too many app and passwords to remember already), is it worth to invest my time on a inferno example for ultra or is it something you don't have interest to expand into?

**Reply by cdoremus** [reply]
The Streaming part of React 18 allows individual UI server-side-rendered components to be asynchronously loaded on a page rather than having the user wait for all page sections to render at once.

**Reply by deckchairlabs** [reply]
We'd like to keep Ultra and it's examples as purely React. I added Preact because it was fairly simple and required very little core code change to support, since it ~matches~ the React API.

**Reply by theArianit** [reply]
ok, got it. Thanks for your replies. I mentioned infernojs because it has the same approach as preact, to be compatible with react, but with some more features out of the box and slightly better performance than react and preact. I think it supports Streaming like React or something similar. anyway, it was just an idea spoken out loud

**Reply by theArianit** [reply]
correction: both, preact and inferno, have experimental streaming.
